working/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,470 @@
+
+    This means that you can't just look at a single type declaration to see what
+    you're allowed to do with it. You have to walk up the hierarchy looking for
+    modifiers. I think it's important for users to be able to quickly tell what
+    they can do with a type just by looking at its declaration, so I don't like
+    this.
+
+*   **No inherited restrictions.** The simplest option is to say that each type
+    gets whatever restrictions you put on it. Since `MySubclass` has no
+    modifiers, it has no restrictions. That's what you wrote, so that's what
+    you get. If that's *not* what you want, then you should put a modifier on
+    it.
+
+    I like the simplicity of this. I think it's consistent with the rest of Dart
+    which is permissive by default. Right now, you can make a class effectively
+    `interface` by giving it only private generative constructors. Since there's
+    no way for a class outside of the library to call one of those constructors,
+    it can't be extended externally. But you could subclass it inside the
+    library with a new class that calls that private generative constructor from
+    its own public one. That subclass is now externally extensible and the
+    language quietly lets you do that.
+
+*   **Trust but verify.** In the above example, it's not clear what the author
+    *intends*. Maybe they deliberately didn't put any modifiers on `MySubclass`
+    because they *want* to re-add the capability that its superclass removed.
+    But maybe they just didn't notice that `NoExtend` removed them, or they
+    forgot to put `interface` on `MySubclass`.
+
+    Since it's not clear what they meant, the language could require them to
+    clarify. If you define a subtype of a type that has removed a capability, we
+    could require you to annotate specifically when you re-add that capability.
+    If you don't intend to re-add a capability, you restate the restriction:
+
+    ```dart
+    interface class NoExtend {}
+    interface class MySubclass extends NoExtend {}
+    ```
+
+    And if you *do* intend to loosen it, you make that explicit by some marker
+    like:
+
+    ```dart
+    interface class NoExtend {}
+    reopen class MySubclass extends NoExtend {}
+    ```
+
+    Here "reopen" means, "I know I didn't put any other modifier here and that
+    means this class has more capabilities than my parent."
+
+    Personally, I think this is probably more modifiers than we want and is
+    more trouble than it's worth. I worry about having to explain to users that
+    a class marked `reopen` means the same thing as a class not marked with it.
+    But I do think it could be useful to offer this as a lint with a metadata
+    annotation for users that are more cautious, like:
+
+    ```dart
+    interface class NoExtend {}
+    @reopen class MySubclass extends NoExtend {}
+    ```
+
+    This proposal takes that option.
+
```

Comment:
Missing the option of *not allowing* you to reopen, and requiring you to repeat the annotation in subclasses inside the library, just as you are required to outside. You can ignore the restrictions, but you cannot remove them. 
If you implement a `base` class, the implementation must still be `base` or `final`, even inside the library.

I know that the options that allow you to reopen also allow you to not do so, so not allowing you to reopen is more restrictive on users. I think it might be worth it, to avoid accidental re-opening.

(I guess the annotation and lint pushes users towards not reopening, so if you never use the annotation/get the lint, you are effectively getting the no-reopening effect. And the compiler can always see whether you reopen or not, and disable any optimizations that don't apply then. I'd just want to consider whether silently disabling optimizations will be a trip-wire for users.)

Reply:
Ah, good point. I'm not in favor of that option, but it's a good one to discuss. Added a section.